### PR TITLE
add representation for multi charge iso pattern

### PR DESCRIPTION
**Pull request by SteffenHeu**

### CODE REVIEW

File: src/main/java/io/github/mzmine/datamodel/impl/MultiChargeStateIsotopePattern.java [hunk 1]
```diff
@@ -248,4 +249,20 @@ public boolean equals(Object o) {
...
+
+  private void onPatternsChanged() {
+    int biggestPatternIndex = -1;
+    int biggestPatternSize = Integer.MIN_VALUE;
+    for (int i = 0; i < patterns.size(); i++) {
+      if (patterns.get(i).getNumberOfDataPoints() > biggestPatternSize) {
+        biggestPatternSize = patterns.get(i).getNumberOfDataPoints();
+        biggestPatternIndex = i;
+      }
+    }
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/datamodel/impl/MultiChargeStateIsotopePattern.java:261]: patterns.stream().max(Comparator.comparingInt(MassSpectrum::getNumberOfDataPoints)).orElse(null)

File: src/main/java/io/github/mzmine/datamodel/impl/MultiChargeStateIsotopePattern.java [hunk 2]
```diff
@@ -248,4 +249,20 @@ public boolean equals(Object o) {
...
+
+    if (biggestPatternIndex != -1) {
+      var preferred = patterns.remove(biggestPatternIndex);
+      addPattern(preferred, true);
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/datamodel/impl/MultiChargeStateIsotopePattern.java:265]: see above. Pattern is added but never removed from the list.

File: src/main/java/io/github/mzmine/modules/visualization/spectra/simplespectra/SpectraVisualizerTab.java [hunk 3]
```diff
@@ -430,39 +433,73 @@ public void loadSinglePeak(Feature peak) {
...
+    Range<Double> searchMZRange = Range.closed(mz - 0.5, mz + 0.5);
+    ScanDataSet scanDataSet = spectrumPlot.getMainScanDataSet();
+    double normalizationFactor = scanDataSet.getHighestIntensity(searchMZRange);
+
+    // If normalization factor is 0, it means there were no data points
+    // in given m/z range. In such case we use the max intensity of
+    // whole scan as normalization factor.
+    if (normalizationFactor == 0) {
+      searchMZRange = Range.atLeast(0.0);
+      normalizationFactor = scanDataSet.getHighestIntensity(searchMZRange);
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/visualization/spectra/simplespectra/SpectraVisualizerTab.java:495]: maybe just get the scan from the dataset and get the basepeak intensity?

File: src/main/java/io/github/mzmine/util/color/ColorUtils.java [hunk 4]
```diff
@@ -30,6 +30,8 @@
 
 public class ColorUtils {
 
+  public static double MIN_REDMEAN_COLOR_DIFF = 65;
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/util/color/ColorUtils.java:33]: javadoc

File: src/main/java/io/github/mzmine/util/color/SimpleColorPalette.java [hunk 5]
```diff
@@ -180,6 +178,19 @@ public synchronized Color getNextColor() {
     return clr;
   }
 
+  public synchronized Color getNextColor(@NotNull final Color exclusion) {
+    final double minDiff = 10d;
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/util/color/SimpleColorPalette.java:182]: jdoc for method?

File: src/main/java/io/github/mzmine/util/color/SimpleColorPalette.java [hunk 6]
```diff
@@ -180,6 +178,19 @@ public synchronized Color getNextColor() {
...
+    final int startNext = next;
+
+    do {
+      var clr = getNextColor();
+      if (ColorUtils.getColorDifference(clr, exclusion) > 10) {
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/util/color/SimpleColorPalette.java:187]: 10 is minDiff?

File: src/test/java/util/ColorUtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    final var yellow = new Color(0.941f, 0.894f, 0.259f, 1f); // yellow
+    final var darkBlue = new Color(0.f, 0.447f, 0.698f, 1f); // blue
+    final var red = new Color(0.835f, 0.369f, 0.f, 1f); // vermillion (darker orange)
+    final var pink = new Color(0.800f, 0.475f, 0.655f, 1f);
+
+    logger.info("" + ColorUtils.getColorDifference(orange, red));
+    logger.info("" + ColorUtils.getColorDifference(brightBlue, darkBlue));
+    logger.info("" + ColorUtils.getColorDifference(darkBlue, Color.BLUE));
+    logger.info("" + ColorUtils.getColorDifference(green, yellow));
+    logger.info("" + ColorUtils.getColorDifference(Color.BLACK, Color.WHITE));
```
Comment (robinschmid) [inline src/test/java/util/ColorUtilsTest.java:54]: turn into test?

File: src/main/java/io/github/mzmine/modules/visualization/spectra/simplespectra/SpectraVisualizerTab.java (hunk 3 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/modules/visualization/spectra/simplespectra/SpectraVisualizerTab.java:495]: this is old functionality which im not changing. fitting to the overall base peak intensity does not make sense cause the pattern shall match the intensity of the representing peak in the scan

File: src/main/java/io/github/mzmine/util/color/SimpleColorPalette.java [hunk 8]
```diff
@@ -180,6 +178,19 @@ public synchronized Color getNextColor() {
...
+        return clr; // this color is different
+      }
+    } while (startNext != next);
```
[the remark this answers is not in the document]
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/util/color/SimpleColorPalette.java:190]: when you looped through the palette once

File: src/main/java/io/github/mzmine/modules/visualization/spectra/simplespectra/SpectraVisualizerTab.java (hunk 3 above)
Comment (robinschmid) [reply src/main/java/io/github/mzmine/modules/visualization/spectra/simplespectra/SpectraVisualizerTab.java:495]: I think at this point in the code there was no signal found in the range and you use the whole mz range. the search range is set to 0

File: src/main/java/io/github/mzmine/util/color/SimpleColorPalette.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (robinschmid) [reply src/main/java/io/github/mzmine/util/color/SimpleColorPalette.java:190]: I see

File: src/main/java/io/github/mzmine/datamodel/impl/MultiChargeStateIsotopePattern.java [hunk 9]
```diff
@@ -112,6 +112,7 @@ public void addPattern(IsotopePattern pattern, boolean setPreferred) {
       patterns.add(0, pattern);
     } else {
       patterns.add(pattern);
+      onPatternsChanged();
```
[the remark this answers is not in the document]
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/datamodel/impl/MultiChargeStateIsotopePattern.java:115]: done

File: src/main/java/io/github/mzmine/modules/visualization/spectra/simplespectra/SpectraVisualizerTab.java (hunk 3 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/modules/visualization/spectra/simplespectra/SpectraVisualizerTab.java:495]: you're right

**Review by robinschmid — APPROVED** [review]
Great thanks for the revisions

**Pull request merged**